### PR TITLE
Implement published test results built in and reduce console chattiness.

### PR DESCRIPTION
**Pull request by BillyONeal**
Built in implementation of https://github.com/microsoft/vcpkg/pull/23477

Before:

```
PS C:\Dev\vcpkg> .\vcpkg.exe install zlib
Computing installation plan...
The following packages will be built and installed:
    zlib[core]:x86-windows -> 1.2.11#13
Detecting compiler hash for triplet x86-windows...
Restored 1 packages from C:\Users\bion\AppData\Local\vcpkg\archives in 43.57 ms. Use --debug to see more details.
Starting package 1/1: zlib:x86-windows
Installing package zlib[core]:x86-windows...
Elapsed time for package zlib:x86-windows: 18.41 ms

Total elapsed time: 2.266 s

The package zlib is compatible with built-in CMake targets:

    find_package(ZLIB REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB)

PS C:\Dev\vcpkg> .\vcpkg.exe remove zlib
The following packages will be removed:
    zlib:x86-windows
Removing package zlib:x86-windows...
PS C:\Dev\vcpkg> .\vcpkg.exe install zlib --no-binarycaching
Computing installation plan...
The following packages will be built and installed:
    zlib[core]:x86-windows -> 1.2.11#13
Detecting compiler hash for triplet x86-windows...
Starting package 1/1: zlib:x86-windows
Building package zlib[core]:x86-windows...
-- Using cached zlib1211.tar.gz.
-- Extracting source C:/Dev/vcpkg-downloads/zlib1211.tar.gz
-- Applying patch cmake_dont_build_more_than_needed.patch
-- Applying patch 0001-Prevent-invalid-inclusions-when-HAVE_-is-set-to-0.patch
-- Applying patch add_debug_postfix_on_mingw.patch
-- Applying patch 0002-android-build-mingw.patch
-- Using source at C:/Dev/vcpkg/buildtrees/zlib/src/1.2.11-e9a093319e.clean
-- Configuring x86-windows
-- Building x86-windows-dbg
-- Building x86-windows-rel
-- Installing: C:/Dev/vcpkg/packages/zlib_x86-windows/share/zlib/vcpkg-cmake-wrapper.cmake
-- Fixing pkgconfig file: C:/Dev/vcpkg/packages/zlib_x86-windows/lib/pkgconfig/zlib.pc
-- Using cached msys-mingw-w64-i686-pkg-config-0.29.2-3-any.pkg.tar.zst.
-- Using cached msys-mingw-w64-i686-libwinpthread-git-9.0.0.6373.5be8fcd83-1-any.pkg.tar.zst.
-- Using msys root at C:/Dev/vcpkg-downloads/tools/msys2/9a1ec3f33446b195
-- Fixing pkgconfig file: C:/Dev/vcpkg/packages/zlib_x86-windows/debug/lib/pkgconfig/zlib.pc
-- Installing: C:/Dev/vcpkg/packages/zlib_x86-windows/share/zlib/copyright
-- Performing post-build validation
-- Performing post-build validation done
Installing package zlib[core]:x86-windows...
Elapsed time for package zlib:x86-windows: 4.83 s

Total elapsed time: 7.009 s

The package zlib is compatible with built-in CMake targets:

    find_package(ZLIB REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB)

PS C:\Dev\vcpkg> .\vcpkg.exe x-set-installed
The following packages will be removed:
    zlib:x86-windows
Starting package 1/1: zlib:x86-windows
Removing package zlib:x86-windows...
Elapsed time for package zlib:x86-windows: 9.93 ms
Restored 0 packages from C:\Users\bion\AppData\Local\vcpkg\archives in 92.5 us. Use --debug to see more details.

Total elapsed time: 14.01 ms

PS C:\Dev\vcpkg> .\vcpkg.exe remove zlib
The following packages are not installed, so not removed:
    zlib:x86-windows
Package zlib:x86-windows is not installed
PS C:\Dev\vcpkg>
```

After:

```
PS C:\Dev\vcpkg> ..\vcpkg-tool\out\build\x64-Debug\vcpkg.exe install zlib
Computing installation plan...
The following packages will be built and installed:
    zlib[core]:x86-windows -> 1.2.11#13
Detecting compiler hash for triplet x86-windows...
Restored 1 packages from C:\Users\bion\AppData\Local\vcpkg\archives in 60.79 ms. Use --debug to see more details.
Installing zlib:x86-windows (1/1)...
Elapsed time for zlib:x86-windows: 20.03 ms

Total elapsed time: 2.662 s

The package zlib is compatible with built-in CMake targets:

    find_package(ZLIB REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB)

PS C:\Dev\vcpkg> ..\vcpkg-tool\out\build\x64-Debug\vcpkg.exe remove zlib
The following packages will be removed:
    zlib:x86-windows
Removing zlib:x86-windows (1/1)...
PS C:\Dev\vcpkg> ..\vcpkg-tool\out\build\x64-Debug\vcpkg.exe install zlib --no-binarycaching
Computing installation plan...
The following packages will be built and installed:
    zlib[core]:x86-windows -> 1.2.11#13
Detecting compiler hash for triplet x86-windows...
Installing zlib:x86-windows (1/1)...
Building zlib[core]:x86-windows...
-- Using cached zlib1211.tar.gz.
-- Cleaning sources at C:/Dev/vcpkg/buildtrees/zlib/src/1.2.11-e9a093319e.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/Dev/vcpkg-downloads/zlib1211.tar.gz
-- Applying patch cmake_dont_build_more_than_needed.patch
-- Applying patch 0001-Prevent-invalid-inclusions-when-HAVE_-is-set-to-0.patch
-- Applying patch add_debug_postfix_on_mingw.patch
-- Applying patch 0002-android-build-mingw.patch
-- Using source at C:/Dev/vcpkg/buildtrees/zlib/src/1.2.11-e9a093319e.clean
-- Configuring x86-windows
-- Building x86-windows-dbg
-- Building x86-windows-rel
-- Installing: C:/Dev/vcpkg/packages/zlib_x86-windows/share/zlib/vcpkg-cmake-wrapper.cmake
-- Fixing pkgconfig file: C:/Dev/vcpkg/packages/zlib_x86-windows/lib/pkgconfig/zlib.pc
-- Using cached msys-mingw-w64-i686-pkg-config-0.29.2-3-any.pkg.tar.zst.
-- Using cached msys-mingw-w64-i686-libwinpthread-git-9.0.0.6373.5be8fcd83-1-any.pkg.tar.zst.
-- Using msys root at C:/Dev/vcpkg-downloads/tools/msys2/9a1ec3f33446b195
-- Fixing pkgconfig file: C:/Dev/vcpkg/packages/zlib_x86-windows/debug/lib/pkgconfig/zlib.pc
-- Installing: C:/Dev/vcpkg/packages/zlib_x86-windows/share/zlib/copyright
-- Performing post-build validation
-- Performing post-build validation done
Elapsed time for zlib:x86-windows: 5.137 s

Total elapsed time: 7.36 s

The package zlib is compatible with built-in CMake targets:

    find_package(ZLIB REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB)

PS C:\Dev\vcpkg> ..\vcpkg-tool\out\build\x64-Debug\vcpkg.exe remove zlib
The following packages will be removed:
    zlib:x86-windows
Removing zlib:x86-windows (1/1)...
PS C:\Dev\vcpkg> ..\vcpkg-tool\out\build\x64-Debug\vcpkg.exe remove zlib
The following packages are not installed, so not removed:
    zlib:x86-windows
Removing zlib:x86-windows (1/1)...
PS C:\Dev\vcpkg>
```

### CODE REVIEW

**Comment by BillyONeal** [comment]
```
	<assembly name="abseil" run-date="2022-03-" run-time="1603:17:" time="25">
		<collection name="x86-windows" time="25">
			<test name="abseil:x64-windows" method="abseil[core]:x64-windows" time="12" result="Pass">
				<traits>
					<trait name="abi_tag" value="a61f543d8b9aede74d38c460244f4b829bcbb226b2fb3d5c421b8b76563c50c8" />
					<trait name="features" value="core" />
					<trait name="owner" value="x64-windows" />
				</traits>
			</test>
			<test name="abseil:x86-windows" method="abseil[core]:x86-windows" time="12" result="Pass">
				<traits>
					<trait name="abi_tag" value="c17d5a9436ead88d57eed554874ab2b714a5a89dee4c7e294ed4b5c5f870bcf6" />
					<trait name="features" value="core" />
					<trait name="owner" value="x86-windows" />
				</traits>
			</test>
		</collection>
	</assembly>
```

File: src/vcpkg/install.cpp [hunk 1]
```diff
@@ -1237,21 +1243,40 @@ namespace vcpkg::Install
         Install::perform_and_exit(args, paths, default_triplet, host_triplet);
     }
 
-    SpecSummary::SpecSummary(const PackageSpec& spec, const Dependencies::InstallPlanAction* action)
-        : spec(spec), build_result{BuildResult::NULLVALUE, nullptr}, action(action)
+    SpecSummary::SpecSummary(const Dependencies::InstallPlanAction* action)
```
Comment (dg0yt) [inline src/vcpkg/install.cpp:1246]: Given that `action` must not be `nullptr`, why isn't it passed with a reference (now) instead of a pointer?

Comment (BillyONeal) [reply src/vcpkg/install.cpp]: Done.

File: include/vcpkg/install.h [hunk 2]
```diff
@@ -25,15 +27,19 @@ namespace vcpkg::Install
 
     struct SpecSummary
```
Comment (BillyONeal) [inline include/vcpkg/install.h:28]: Long term, I think SpecSummary should be "freestanding" and not require the action plan that produced it to be held "alive" in order to be meaningful but that turned out to be more effort than I was willing to spend here. This at least goes somewhat in that direction.

**Comment by ras0219-msft** [comment]
Message diff, since there's much more unchanged text than changed text.

```diff
diff --git a/t1.txt b/t2.txt
index cb2874848..91f7d2012 100644
--- a/t1.txt
+++ b/t2.txt
@@ -4,9 +4,8 @@ The following packages will be built and installed:
     zlib[core]:x86-windows -> 1.2.11#13
 Detecting compiler hash for triplet x86-windows...
 Restored 1 packages from C:\Users\bion\AppData\Local\vcpkg\archives in x ms. Use --debug to see more details.
-Starting package 1/1: zlib:x86-windows
-Installing package zlib[core]:x86-windows...
-Elapsed time for package zlib:x86-windows: x ms
+Installing zlib:x86-windows (1/1)...
+Elapsed time for zlib:x86-windows: x ms
 
 Total elapsed time: x s
 
@@ -18,14 +17,14 @@ The package zlib is compatible with built-in CMake targets:
 PS C:\Dev\vcpkg> .\vcpkg.exe remove zlib
 The following packages will be removed:
     zlib:x86-windows
-Removing package zlib:x86-windows...
+Removing zlib:x86-windows (1/1)...
 PS C:\Dev\vcpkg> .\vcpkg.exe install zlib --no-binarycaching
 Computing installation plan...
 The following packages will be built and installed:
     zlib[core]:x86-windows -> 1.2.11#13
 Detecting compiler hash for triplet x86-windows...
-Starting package 1/1: zlib:x86-windows
-Building package zlib[core]:x86-windows...
+Installing zlib:x86-windows (1/1)...
+Building zlib[core]:x86-windows...
 -- Using cached zlib1211.tar.gz.
 -- Extracting source C:/Dev/vcpkg-downloads/zlib1211.tar.gz
 -- Applying patch cmake_dont_build_more_than_needed.patch
@@ -45,8 +44,7 @@ Building package zlib[core]:x86-windows...
 -- Installing: C:/Dev/vcpkg/packages/zlib_x86-windows/share/zlib/copyright
 -- Performing post-build validation
 -- Performing post-build validation done
-Installing package zlib[core]:x86-windows...
-Elapsed time for package zlib:x86-windows: x s
+Elapsed time for zlib:x86-windows: x s
 
 Total elapsed time: x s
 
@@ -55,18 +53,12 @@ The package zlib is compatible with built-in CMake targets:
     find_package(ZLIB REQUIRED)
     target_link_libraries(main PRIVATE ZLIB::ZLIB)
 
-PS C:\Dev\vcpkg> .\vcpkg.exe x-set-installed
+PS C:\Dev\vcpkg> .\vcpkg.exe remove zlib
 The following packages will be removed:
     zlib:x86-windows
-Starting package 1/1: zlib:x86-windows
-Removing package zlib:x86-windows...
-Elapsed time for package zlib:x86-windows: x ms
-Restored 0 packages from C:\Users\bion\AppData\Local\vcpkg\archives in x us. Use --debug to see more details.
-
-Total elapsed time: x ms
-
+Removing zlib:x86-windows (1/1)...
 PS C:\Dev\vcpkg> .\vcpkg.exe remove zlib
 The following packages are not installed, so not removed:
     zlib:x86-windows
-Package zlib:x86-windows is not installed
+Removing zlib:x86-windows (1/1)...
 PS C:\Dev\vcpkg>
\ No newline at end of file
```

File: src/vcpkg/remove.cpp [hunk 3]
```diff
@@ -305,8 +300,11 @@ namespace vcpkg::Remove
             Checks::exit_success(VCPKG_LINE_INFO);
         }
 
-        for (const RemovePlanAction& action : remove_plan)
+        // note that we try to "remove" things that aren't installed to trigger purge actions
+        for (std::size_t idx = 0; idx < remove_plan.size(); ++idx)
         {
+            const RemovePlanAction& action = remove_plan[idx];
+            vcpkg::printf("Removing %s (%zu/%zu)...\n", action.spec.to_string(), idx + 1, remove_plan.size());
```
Comment (strega-nil-ms) [inline src/vcpkg/remove.cpp:307]: This should probably be localized.

File: src/vcpkg/install.cpp [hunk 4]
```diff
@@ -447,18 +448,28 @@ namespace vcpkg::Install
...
+            current_summary = &results.back();
+            vcpkg::printf("Removing %s (%zd/%zd)...\n", action.spec.to_string(), action_index, action_count);
         }
 
         ~TrackedPackageInstallGuard()
         {
             current_summary->timing = build_timer.elapsed();
             vcpkg::printf(
-                "Elapsed time for package %s: %s\n", current_summary->spec.to_string(), current_summary->timing);
+                "Elapsed time for %s: %s\n", current_summary->get_spec().to_string(), current_summary->timing);
```
Comment (strega-nil-ms) [inline src/vcpkg/install.cpp:472]: same here

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
Please localize while you're touching these files; approved post that

**Review by strega-nil-ms — APPROVED** [review]

**Review by ras0219-msft — APPROVED** [review]
I converted the xUnit generator to use our XmlSerializer machinery.

I also pushed localization for all the touched messages.

File: src/vcpkg/base/xmlserializer.cpp [hunk 5]
```diff
@@ -90,6 +90,16 @@ namespace vcpkg
         }
         return *this;
     }
+    XmlSerializer& XmlSerializer::cdata(StringView sv)
```
Comment (BillyONeal) [inline src/vcpkg/base/xmlserializer.cpp:93]: @ras0219-msft I think this CDATA thing was a workaround for not having real XmlSerializer available at the time; should this go through data() instead now that you have proper escapes implemented?

File: src/vcpkg/remove.cpp [hunk 6]
```diff
@@ -305,8 +302,12 @@ namespace vcpkg::Remove
             Checks::exit_success(VCPKG_LINE_INFO);
         }
 
-        for (const RemovePlanAction& action : remove_plan)
+        // note that we try to "remove" things that aren't installed to trigger purge actions
+        for (std::size_t idx = 0; idx < remove_plan.size(); ++idx)
         {
+            const RemovePlanAction& action = remove_plan[idx];
+            msg::println(msg::format(msgRemovingPackage, msg::spec = action.spec)
+                             .append_fmt_raw(" ({}/{})...", idx + 1, remove_plan.size()));
```
Comment (BillyONeal) [inline src/vcpkg/remove.cpp:310]: I see you added `append_fmt_raw` now; is this (x/y) notation common enough that we should add an explicit fn for it? (There are at least 2-3 such examples touched in this PR alone but I don't have a barometer on how many others there are)

**Review by CryptoGodfatherVA23 — APPROVED** [review]

File: src/vcpkg/base/xmlserializer.cpp (hunk 5 above)
Comment (ras0219-msft) [reply src/vcpkg/base/xmlserializer.cpp:93]: The spec actually says "as a CDATA block": https://xunit.net/docs/format-xml-v2

(for the &lt;message> child of a &lt;failure> node)
> The failure message as a CDATA block.

File: src/vcpkg/remove.cpp (hunk 6 above)
Comment (ras0219-msft) [reply src/vcpkg/remove.cpp:310]: These 2-3 are the only ones I've touched recently. I think that refactor can be done separately if/when it's deemed needed.

File: src/vcpkg/base/xmlserializer.cpp (hunk 5 above)
Comment (BillyONeal) [reply src/vcpkg/base/xmlserializer.cpp:93]: WOW!

**Comment by autoantwort** [comment]
Hm I dislike that the `x/y` status is now displayed at another location for every package. This makes tracking the installation process harder.

**Comment by BillyONeal** [comment]
> Hm I dislike that the `x/y` status is now displayed at another location for every package. This makes tracking the installation process harder.

I don’t feel strongly about where  that is, only that the relatively useless “starting package” line is byebye

**Comment by dg0yt** [comment]
Having `[x/y]` first helps a lot when reading logs. For illustration, this is Ninja output:
~~~
[1/2] C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x64\cl.exe   /TP -DUNICODE -D_UNICODE -ID:\installed\x64-windows-static\lib\mswu -ID:\installed\x64-windows-static\include /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /GR /EHsc /MP  /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1  -MTd /showIncludes /FoCMakeFiles\main.dir\popup.cpp.obj /FdCMakeFiles\main.dir\ /FS -c D:\installed\x64-windows-static\share\wxwidgets\example\popup.cpp
[2/2] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.22.2-windows\cmake-3.22.2-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=CMakeFiles\main.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x64\link.exe  CMakeFiles\main.dir\popup.cpp.obj  /out:main.exe /implib:main.lib /pdb:main.pdb /version:0.0 /machine:x64 /nologo    /debug /INCREMENTAL /subsystem:console  D:\installed\x64-windows-static\debug\lib\wxmsw31ud_core.lib  D:\installed\x64-windows-static\debug\lib\wxbase31ud.lib  D:\installed\x64-windows-static\debug\lib\wxregexud.lib  winmm.lib  comctl32.lib  uuid.lib  oleacc.lib  uxtheme.lib  rpcrt4.lib  shlwapi.lib  version.lib  wsock32.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
~~~

**Comment by BillyONeal** [comment]
> Having `[x/y]` first helps a lot when reading logs.

Sure, that can make sense. Given that we didn't do that before this change though I'm not sure what's being seen as the regression?

**Pull request merged**